test/integration/replicaset/replicaset_test.go

```diff
@@ -268,250 +493,454 @@ func TestAdoption(t *testing.T) {
+	if newRS.Status.AvailableReplicas != 1 {
+		t.Fatalf("Unexpected .Status.AvailableReplicas: Expected 1, saw %d", newRS.Status.AvailableReplicas)
 	}
-	close(stopCh)
 }
 
-func TestUpdateLabelToBeAdopted(t *testing.T) {
-	// We have pod1, pod2 and rs. rs.spec.replicas=1. At first rs.Selector
-	// matches pod1 only; change pod2's labels to be matching. Verify the RS
-	// controller adopts pod2 and delete one of them, so there is only 1 pod
-	// left.
-	s, closeFn, rm, informers, clientSet := rmSetup(t)
+func TestRSScaleSubresource(t *testing.T) {
+	s, closeFn, rm, informers, c := rmSetup(t)
 	defer closeFn()
-	ns := framework.CreateTestingNamespace("rs-update-label-to-be-adopted", s, t)
+	ns := framework.CreateTestingNamespace("test-rs-scale-subresource", s, t)
 	defer framework.DeleteTestingNamespace(ns, s, t)
+	stopCh := runControllerAndInformers(t, rm, informers, 0)
+	defer close(stopCh)
+
 	rs := newRS("rs", ns.Name, 1)
-	// let rs's selector only matches pod1
-	rs.Spec.Selector.MatchLabels["uniqueKey"] = "1"
-	rs.Spec.Template.Labels["uniqueKey"] = "1"
-	pod1 := newMatchingPod("pod1", ns.Name)
-	pod1.Labels["uniqueKey"] = "1"
-	pod2 := newMatchingPod("pod2", ns.Name)
-	pod2.Labels["uniqueKey"] = "2"
-	createRSsPods(t, clientSet, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{pod1, pod2}, ns.Name)
+	rss, _ := createRSsPods(t, c, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{})
+	rs = rss[0]
+	waitRSStable(t, c, rs)
+
+	// Use scale subresource to scale up .Spec.Replicas to 3
+	testScalingUsingScaleSubresource(t, c, rs, 3)
+	// Use the scale subresource to scale down .Spec.Replicas to 0
+	testScalingUsingScaleSubresource(t, c, rs, 0)
+}
 
-	stopCh := make(chan struct{})
-	informers.Start(stopCh)
-	go rm.Run(5, stopCh)
-	waitRSStable(t, clientSet, rs, ns.Name)
+func TestExtraPodsAdoptionAndDeletion(t *testing.T) {
+	s, closeFn, rm, informers, c := rmSetup(t)
+	defer closeFn()
+	ns := framework.CreateTestingNamespace("test-extra-pods-adoption-and-deletion", s, t)
+	defer framework.DeleteTestingNamespace(ns, s, t)
 
-	// change the rs's selector to match both pods
-	patch := `{"metadata":{"labels":{"uniqueKey":"1"}}}`
-	podClient := clientSet.Core().Pods(ns.Name)
-	pod2, err := podClient.Patch(pod2.Name, types.StrategicMergePatchType, []byte(patch))
-	if err != nil {
-		t.Fatalf("Failed to patch pod2: %v", err)
-	}
-	t.Logf("patched pod2 = %#v", pod2)
-	// wait for the rs to select both pods and delete one of them
-	if err := wait.Poll(10*time.Second, 60*time.Second, func() (bool, error) {
-		return verifyRemainingObjects(t, clientSet, ns.Name, 1, 1)
+	rs := newRS("rs", ns.Name, 2)
+	// Create 3 pods, RS should adopt only 2 of them
+	podList := []*v1.Pod{}
+	for i := 0; i < 3; i++ {
+		pod := newMatchingPod(fmt.Sprintf("pod-%d", i+1), ns.Name)
+		pod.Labels = labelMap()
+		podList = append(podList, pod)
+	}
+	rss, _ := createRSsPods(t, c, []*v1beta1.ReplicaSet{rs}, podList)
+	rs = rss[0]
+
+	// Verify there are 3 pods before starting RS controller
```

Comment:
This is already verified for you now that `runControllerAndInformers` calls `waitToObservePods`.

Reply:
Agreed. The check is redundant.